azkaban-web-server/src/main/java/azkaban/imagemgmt/services/ImageRampRuleServiceImpl.java

```diff
@@ -0,0 +1,93 @@
+package azkaban.imagemgmt.services;
+
+import azkaban.imagemgmt.daos.ImageTypeDao;
+import azkaban.imagemgmt.daos.ImageVersionDao;
+import azkaban.imagemgmt.daos.RampRuleDao;
+import azkaban.imagemgmt.dto.ImageRampRuleRequestDTO;
+import azkaban.imagemgmt.exception.ErrorCode;
+import azkaban.imagemgmt.exception.ImageMgmtInvalidInputException;
+import azkaban.imagemgmt.exception.ImageMgmtInvalidPermissionException;
+import azkaban.imagemgmt.models.ImageOwnership;
+import azkaban.imagemgmt.models.ImageRampRule;
+import azkaban.imagemgmt.models.ImageType;
+import azkaban.user.UserManager;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.log4j.Logger;
+
+@Singleton
+public class ImageRampRuleServiceImpl implements ImageRampRuleService {
+
+  public static final Logger log = Logger.getLogger(ImageRampRuleServiceImpl.class);
+
+  private final RampRuleDao rampRuleDao;
+  private final ImageTypeDao imageTypeDao;
+  private final ImageVersionDao imageVersionDao;
+  private final UserManager userManager;
+
+  @Inject
+  public ImageRampRuleServiceImpl(RampRuleDao rampRuleDao,
+                                  ImageTypeDao imageTypeDao,
+                                  ImageVersionDao imageVersionDao,
+                                  UserManager userManager) {
+    this.rampRuleDao = rampRuleDao;
+    this.imageTypeDao = imageTypeDao;
+    this.imageVersionDao = imageVersionDao;
+    this.userManager = userManager;
+  }
+
+  @Override
+  public void createRule(ImageRampRuleRequestDTO rampRuleRequest, String ldapUser, boolean isAzkabanAdmin){
+    // validate image_name and image_version
+    final ImageType imageType = imageTypeDao
+      .getImageTypeByName(rampRuleRequest.getImageName())
+        .orElseThrow(() -> new ImageMgmtInvalidInputException(ErrorCode.NOT_FOUND, String.format("Unable to"
+            + " fetch image type metadata. Invalid image type: %s.", rampRuleRequest.getImageName())));
+    if (!this.imageVersionDao.isInvalidVersion(rampRuleRequest.getImageName(), rampRuleRequest.getImageVersion())) {
+      throw new ImageMgmtInvalidInputException(ErrorCode.NOT_FOUND, String.format(
+          "Unable to fetch image version metadata. Invalid image version: %s.", rampRuleRequest.getImageVersion()));
+    }
+    boolean isAuthorized = isAzkabanAdmin;
```

Comment:
Can you reuse hasImageManagementPermission() method to tell if a user is authorized or not?
Also put this authorization into *Servlet: e.g. see how we do it in ImageRampupServlet.

Reply:
I saw that code. The reason I am not using that is I think this is duplicated code. I need to getOwnerships too instead of simply check if this is authorized. I setup ownership of the rule instead image ownership directly to handle the different cases, for example hp flows; and later API would only rely on rule owner to authorize instead of looking into image owner.
If you still think reusing PermissionManager is needed. I will probably add validatedPermissionAndGetImageOwnerships
validatedPermissionAndGetHPFlowOwnerships
hasPermissionOnRule
as new method,
change hasPermission to hasPermissionOnImage

Reply:
It makes sense to me. Thanks for the clarification!